File: test/react-native/features/fixtures/scenario-launcher/android/src/main/java/com/bugsnag/reactnative/test/scenarios/GroupingDiscriminatorNativeScenario.kt
package com.reactnative.scenarios

import android.content.Context
import com.bugsnag.android.Bugsnag
import com.facebook.react.bridge.Promise

class GroupingDiscriminatorNativeScenario(context: Context): Scenario(context) {

    override fun run(promise: Promise) {
        val exception = RuntimeException("GroupingDiscriminatorScenarioNative")
        Bugsnag.notify(exception)
        Thread.sleep(500)
        Bugsnag.setGroupingDiscriminator("grouping-discriminator-from-native")
        // JS layer will be automatically notified via the BugsnagReactNativeEmitter
        // when setGroupingDiscriminator is called, which triggers the client observer
    }

    override fun runSync(): Boolean {
        val exception = RuntimeException("GroupingDiscriminatorScenarioNative")
        Bugsnag.notify(exception)
        Thread.sleep(500)
        Bugsnag.setGroupingDiscriminator("grouping-discriminator-from-native")
        return true
    }
}

File: test/react-native/features/fixtures/scenario-launcher/ios/Scenarios/GroupingDiscriminatorNativeScenario.m
#import "GroupingDiscriminatorNativeScenario.h"

@implementation GroupingDiscriminatorNativeScenario

- (void)run: (RCTPromiseResolveBlock)resolve
     reject:(RCTPromiseRejectBlock)reject {
  NSException *exception = [[NSException alloc] initWithName:@"NSException" reason:@"GroupingDiscriminatorScenarioNative" userInfo:nil];
  [Bugsnag notify:exception];
  [NSThread sleepForTimeInterval:0.5];
  [Bugsnag setGroupingDiscriminator:@"grouping-discriminator-from-native"];
  // JS layer will be automatically notified via the BugsnagReactNativeEmitter
  // when setGroupingDiscriminator is called, which triggers BSGClientObserverUpdateGroupingDiscriminator
}

@end
